### PR TITLE
New: Decouple configuration unit tests from main suite

### PR DESCRIPTION
**Pull request by dsankouski**
Configuration unit tests affects whole TestNG tests. That is, when running 1 random test from IDE, (@(Before|After)Suite) configuration also run and fail (because no related configuration test was run).

Solution introduced to run dedicated programmatically created test suite inside configuration test, and analyse results, thus decoupling (@(Before|After)Suite) configuration from main suite.

- [x] Add test case(s) - not applicable
- [x] Update `CHANGES.txt`

### CODE REVIEW

File: testng-core/src/test/java/test/configuration/ConfigurationTest.java [hunk 1]
```diff
@@ -1,98 +1,30 @@
...
   @Test
-  public void verify() {
-    assert m_beforeSuite : "beforeSuite should have run";
-    assert m_beforeClass : "beforeClass have run";
-    assert m_beforeMethod : "beforeMethod should have run";
-    assert !m_afterSuite : "afterSuite shouldn't have run";
-    assert !m_afterClass : "afterClass shouldn't have run";
-    assert !m_afterMethod : "afterMethod shouldn't have run";
+  public void testSuite() {
+    testConfiguration(SuiteTestSample.class, VerifySuiteTestSample.class);
```
Comment (juherr) [inline testng-core/src/test/java/test/configuration/ConfigurationTest.java:28]: In fact, `VerifySuiteTestSample` is the expected test code for `SuiteTestSample`.

File: testng-core/src/test/java/test/configuration/sample/ExternalConfigurationClass.java [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package test.configuration;
+package test.configuration.sample;
```
Comment (juherr) [inline testng-core/src/test/java/test/configuration/sample/ExternalConfigurationClass.java:1]: By convention, suffix the class name by `Sample` too

File: testng-core/src/test/java/test/configuration/sample/ConfigurationTestSample.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  private boolean m_beforeSuite = false;
+  private boolean m_afterSuite = false;
+  private boolean m_beforeClass = false;
+  private boolean m_afterClass = false;
+  private boolean m_beforeMethod = false;
+  private boolean m_afterMethod = false;
+
+  @BeforeSuite
+  public void beforeSuite() {
+    assert !m_afterSuite : "afterSuite shouldn't have run";
```
[the remark this answers is not in the document]
Comment (juherr) [reply testng-core/src/test/java/test/configuration/sample/ConfigurationTestSample.java:26]: See `ExternalConfigurationClass`

File: testng-core/src/test/java/test/configuration/sample/SuiteTestSample.java [hunk 4]
```diff
@@ -1,20 +1,15 @@
-package test.configuration;
+package test.configuration.sample;
 
 import org.testng.annotations.BeforeSuite;
 import org.testng.annotations.Test;
+import test.configuration.BaseSuiteTest;
 
-public class SuiteTest extends BaseSuiteTest {
+public class SuiteTestSample extends BaseSuiteTest {
```
Comment (juherr) [inline testng-core/src/test/java/test/configuration/sample/SuiteTestSample.java:7]: A Sample must not extend `BaseSuiteTest `

**Review by juherr — CHANGES_REQUESTED** [review]
Thanks for the PR. The idea is good and I hope we will be able to merge it asap :)

File: testng-core/src/test/java/test/configuration/sample/SuiteTestSample.java [hunk 5]
```diff
@@ -1,20 +1,15 @@
...
   @BeforeSuite(dependsOnMethods = {"beforeSuiteParent"})
   public void beforeSuiteChild() {
     m_order.add(2);
   }
 
-  //  @AfterSuite(dependsOnMethods={"afterSuiteParent"})
```
[the remark this answers is not in the document]
Comment (dsankouski) [reply testng-core/src/test/java/test/configuration/sample/SuiteTestSample.java:12]: yes. The question is: does `dependsOnMethods` takes advantage over default inheritance order?

File: testng-core/src/test/java/test/configuration/sample/ConfigurationTestSample.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.AfterSuite;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Test;
+
+/**
```
[the remark this answers is not in the document]
Comment (dsankouski) [reply testng-core/src/test/java/test/configuration/sample/ConfigurationTestSample.java]: removed

File: testng-core/src/test/java/test/configuration/sample/ConfigurationTestSample.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dsankouski) [reply testng-core/src/test/java/test/configuration/sample/ConfigurationTestSample.java]: done

File: testng-core/src/test/java/test/factory/Factory2Test.java [hunk 7]
```diff
@@ -1,15 +1,13 @@
...
- */
-public class Factory2Test {
-
-  @Factory()
-  public Object[] createObjects() {
-    return new Object[] {new Factory2Sample(), new Factory2Sample()};
+public class Factory2Test extends ConfigurationBaseTest {
+  @Test
+  public void testFactoryCorrectlyInterleaved() {
+    testConfiguration(Factory2Sample.class, Factory2TestSample.class);
```
[the remark this answers is not in the document]
Comment (dsankouski) [reply testng-core/src/test/java/test/factory/Factory2Test.java]: removed

File: testng-core/src/test/java/test/configuration/sample/SuiteTestSample.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dsankouski) [reply testng-core/src/test/java/test/configuration/sample/SuiteTestSample.java:12]: I assume yes for now, it seems more logical to me

File: testng-core/src/test/java/test/configuration/sample/VerifySuiteTestSample.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.util.Arrays;
+import org.testng.Assert;
+import org.testng.annotations.AfterSuite;
+
+public class VerifySuiteTestSample {
+
+  @AfterSuite
+  public void verify() {
+    Assert.assertEquals(Arrays.asList(1, 2, 3), SuiteTestSample.m_order);
```
[the remark this answers is not in the document]
Comment (dsankouski) [reply testng-core/src/test/java/test/configuration/sample/VerifySuiteTestSample.java]: moved

File: testng-core/src/test/java/test/configuration/ConfigurationTest.java (hunk 1 above)
Comment (dsankouski) [reply testng-core/src/test/java/test/configuration/ConfigurationTest.java]: VerifySuiteTestSample.class removed

File: testng-core/src/test/java/test/configuration/sample/ExternalConfigurationClass.java (hunk 2 above)
Comment (dsankouski) [reply testng-core/src/test/java/test/configuration/sample/ExternalConfigurationClass.java]: done

File: testng-core/src/test/java/test/configuration/sample/SuiteTestSample.java (hunk 4 above)
Comment (dsankouski) [reply testng-core/src/test/java/test/configuration/sample/SuiteTestSample.java]: BaseSuiteTest moved to samples

**Review by juherr — APPROVED** [review]

**Comment by juherr** [comment]
@dsankouski Could you fix the conflict?

@krmahadevan LGTM, could you check too?

**Review by krmahadevan — APPROVED** [review]

**Pull request merged**